pennylane/tape/tape.py

```diff
@@ -459,60 +448,23 @@ def _process_queue(self):
         self._prep = []
         self._ops = []
         self._measurements = []
-        self._output_dim = 0
+        list_order = {"_prep": 0, "_ops": 1, "_measurements": 2}
+        current_list = "_prep"
 
         for obj, info in self._queue.items():
 
-            if isinstance(obj, QuantumTape):
-                self._ops.append(obj)
-
-            elif isinstance(obj, qml.operation.Operation) and not info.get("owner", False):
-                # operation objects with no owners
-
-                if self._measurements:
+            if "owner" not in info and getattr(obj, "_queue_category", None) is not None:
+                if list_order[obj._queue_category] > list_order[current_list]:
+                    current_list = obj._queue_category
+                elif list_order[obj._queue_category] < list_order[current_list]:
                     raise ValueError(
-                        f"Quantum operation {obj} must occur prior to any measurements."
+                        f"Object {obj} must occur prior to {current_list}. "
```

Comment:
This reduces code, but the user gets a very funky error message :)

Reply:
Maybe this (with the suggestion above)?

```suggestion
                        msg[current_list](obj)
```